config/configsource/configsource.go

```diff
@@ -0,0 +1,267 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package configsource
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+
+	"github.com/spf13/viper"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+)
+
+const (
+	// ConfigSourcePrefix is used to identified a configuration source invocation.
+	ConfigSourcePrefix = "$"
+)
+
+// ApplyConfigSourcesParams holds the parameters for injecting data from the
+// given configuration sources into a configuration.
+type ApplyConfigSourcesParams struct {
+	// Set of configuration sources available to inject data into the configuration.
+	ConfigSources map[string]component.ConfigSource
+	// MaxRecursionDepth limits the maximum number of times that a configuration source
+	// can inject another into the config.
+	MaxRecursionDepth uint
+	// LoadCount tracks the number of the configuration load session.
+	LoadCount int
+}
+
+// ApplyConfigSources takes a viper object with the configuration to which the
+// the configuration sources will be applied and the resulting configuration is
+// returned as a separated object.
+func ApplyConfigSources(ctx context.Context, v *viper.Viper, params ApplyConfigSourcesParams) (*viper.Viper, error) {
```

Comment:
Please avoid exposing viper for the moment? No requirement to fix now since this is experimental, but sometimes soon this needs to change since we want to move away from exposing Viper in the public APIs, it is ok to use it internally.

Reply:
To avoid using Viper as an alternate should we perhaps accept the config as a string and return as a string?

Reply:
I opted to expose viper so the function could be called before passing the viper to `config.Load` that unfortunately receives it as a parameter. What we need is a map of keys to their config values. I will look into alternatives in follow-up PRs. Keeping as viper for the time being.

Reply:
Resolving since this is not part of the PR anymore.